---
title: "Sex-related analysis"
subtitle: "Using Dream"
author:  |
 | Katia de Paiva Lopes 
 | Raj Lab
 | Department of Neuroscience
 | Icahn School of Medicine at Mount Sinai
 | NYC, New York
date: "`r Sys.Date()`"
output: 
 rmarkdown::html_document:
   theme: united
   highlight: tango
   code_folding: hide
   toc: true
   toc_float: true
   df_print: paged
   smooth_scroll: true
   number_sections: false
   self_contained: true  
---

> Dream tool | coef = sex   

```{r clean.variables, echo=FALSE}
#This command clean all variables. BE CAREFULL!!! 
rm(list = setdiff(ls(), lsf.str()))
```

```{r load.packages, echo=FALSE, message=FALSE, results='hide'}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
if(!require("ggeasy")) BiocManager::install("ggeasy"); library("ggeasy")
if(!require("ggsci")) BiocManager::install("ggsci"); library("ggsci")
if(!require("ggplot2")) BiocManager::install("ggplot2"); library("ggplot2")
if(!require("factoextra")) BiocManager::install("factoextra"); library("factoextra")
if(!require("grid")) BiocManager::install("grid"); library("grid")
if(!require("variancePartition")) BiocManager::install("variancePartition"); library("variancePartition")
if(!require("BiocParallel")) install.packages("BiocParallel"); library("BiocParallel")
if(!require("edgeR")) install.packages("edgeR"); library("edgeR")
if(!require("statmod")) install.packages("statmod"); library("statmod")
if(!require("gplots")) install.packages("gplots"); library("gplots")
if(!require("ggpubr")) install.packages("ggpubr"); library("ggpubr")
library(readxl)
library(pheatmap)
library(reshape2)
library(kableExtra)
library(RColorBrewer)
require(gridExtra)
library(ggfortify)
library('doParallel')
library(dplyr)
library(tidyr)

knitr::opts_chunk$set( 
  warning=FALSE,
  message=FALSE,
  results = 'asis',
  error = FALSE,
  tidy = FALSE,
  fig.show = "hold")

#register(MulticoreParam(workers = 2), default = TRUE)
```

```{r Helper Functions, echo=FALSE}
createDT <- function(DF, caption="", scrollY=500){
  data <- DT::datatable(DF, caption=caption,
    extensions =  'Buttons',
    options = list( dom = 'Bfrtip', 
                    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), 
                    scrollY = scrollY, scrollX=T, scrollCollapse = T, paging = F,  
                      columnDefs = list(list(className = 'dt-center', targets = "_all"))
    )
  ) 
   return(data)
}
```

```{r folders, echo=TRUE}
expression_dir = "~/ad-omics_hydra/microglia_omics/expression_tables/added_pilot_314s/expr_4brain_regions/"
work_plots = "~/pd-omics/katia/scripts/GitHub_scripts/glia_omics/3rd_pass_mic_255s/sex_related/"
metadata_path = "~/pd-omics/katia/Microglia/mic_255s/metadata_files/"
```

```{r input.data, echo=TRUE}
load(paste0(expression_dir, "Expression_filt_255s.Rdata")) 
# dim(genes_counts_exp_3rd) # 19376   255

load(paste0(metadata_path, "metadata_255filt_eng_29jun2020.Rdata"))
# str(metadata3rd_pass)

metadata <- metadata3rd_pass
genes_counts_exp <- genes_counts_exp_3rd 
```

## Sex 
```{r sex.count, echo=TRUE}
sexByDonor = unique(metadata[,c("donor_id", "sex")])
#createDT(sexByDonor)
as.data.frame(t(as.matrix(unclass(  table(sexByDonor$sex, useNA = "ifany"))))) %>%
 kable(row.names = F) %>%
 kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

## Remove the X and Y genes 
I've removed 619 genes in our dataset, from x and y chromosomes. 
```{r chr, echo=TRUE}
position = read.table("/sc/hydra/projects/PBG/REFERENCES/GRCh38/Gencode/release_30/gencode.v30.primary_assembly.annotation.reflat", stringsAsFactors = F, header = F)
position = position[, c(1,3)]
colnames(position) = c("ensembl", "chr")
y_ens = position$ensembl[position$chr %in% "chrY"] 
x_ens = position$ensembl[position$chr %in% "chrX"] 
genes_counts_nosex = genes_counts_exp[!rownames(genes_counts_exp) %in% c(x_ens, y_ens), ]
```

## Dream analysis {.tabset .tabset-fade .tabset-pills} 
```{r dream, echo=TRUE}
params = BiocParallel::MulticoreParam(workers=20, progressbar=T)
register(params)
registerDoParallel(20)

metadata$cause_of_death_categories[metadata$cause_of_death_categories %in% NA] <- "Other"
#table(metadata$cause_of_death_categories)

metadata$C1 = metadata$C1 %>% replace_na(median(metadata$C1, na.rm = T))
metadata$C2 = metadata$C2 %>% replace_na(median(metadata$C2, na.rm = T))
metadata$C3 = metadata$C3 %>% replace_na(median(metadata$C3, na.rm = T))
metadata$C4 = metadata$C4 %>% replace_na(median(metadata$C4, na.rm = T))

# Matching the names from the DE analysis to use the same code 
genes_counts4deg = genes_counts_nosex
metadata4deg = metadata
# all(colnames(genes_counts_exp) == metadata$donor_tissue) # Check the order of columns - TRUE 

# The dream model operates directly on the results of voom. 
# The only change compared to the standard limma workflow is to replace lmFit with dream. 

# Check variance partition version 
# packageVersion("variancePartition")  # Must be 1.17.7

# The variable to be tested should be a fixed effect
form <- ~ sex + (1|donor_id) + age + tissue + (1|cause_of_death_categories) + C1 + C2 + C3 + C4 + picard_pct_mrna_bases + picard_summed_median + picard_pct_ribosomal_bases

# estimate weights using linear mixed model of dream
vobjDream = suppressWarnings( voomWithDreamWeights( genes_counts4deg, form, metadata4deg ) ) # supressing messages because of Biocparallel was generating a lot of messages  
 
# Fit the dream model on each gene
# By default, uses the Satterthwaite approximation for the hypothesis test
fitmm = suppressWarnings (dream( vobjDream, form, metadata4deg )) 

# Examine design matrix
# createDT(fitmm$design, 3)

res_dream <- data.frame(topTable(fitmm, coef='sexm', 
                                 number=nrow(genes_counts4deg), sort.by = "p"), check.names = F)

```

The t-statistics are not directly comparable since they have different degrees of freedom. In order to be able to compare test statistics, we report z.std which is the p-value transformed into a signed z-score. This can be used for downstream analysis.

## DE genes 15% 
```{r dream.deg015, echo=TRUE, results='asis'}
length(which(res_dream$adj.P.Val<0.15))
```

## DE genes 10%
```{r dream.deg0001, echo=TRUE, results='asis'}
length(which(res_dream$adj.P.Val<0.10))
```

## DE genes 5% 
```{r dream.deg005, echo=TRUE, results='asis'}
length(which(res_dream$adj.P.Val<0.05))
```

## Plots from Dream {.tabset .tabset-fade .tabset-pills}
### P Value Distribution {.tabset .tabset-fade .tabset-pills}
```{r design0.pvalue, echo=TRUE, res=300}
res = res_dream
p = ggplot(res, aes(P.Value))
p + geom_density(color="darkblue", fill="lightblue") +
  theme_classic() +
  ggtitle("FDR Distribution")
```

### Fold Change Distribution {.tabset .tabset-fade .tabset-pills}
```{r design0.FC, echo=TRUE, dpi = 300}
p = ggplot(res, aes(logFC))
p + geom_density(color = "darkblue", fill = "lightblue") + 
  theme_classic() +
  ggtitle("Fold Change Distribution")
```

### MA Plot {.tabset .tabset-fade .tabset-pills}
```{r design0.MA, echo=TRUE, dpi = 300}
plot.data = res
plot.data$id = rownames(plot.data)
data = data.frame(plot.data)
data$P.Value = -log10(data$P.Value)
data$fifteen = as.factor(abs(data$adj.P.Val < 0.05))

ma = ggplot(data, aes(AveExpr, logFC, color = fifteen))
ma + geom_point() +
  scale_color_manual(values = c("black", "red"), labels = c ("> 0.05", "< 0.05")) +
  labs(title = "MA plot", color = "labels") +
  theme_classic()
  #theme(plot.title = element_text(hjust = 0.5)) + ylim (-10,10) + xlim(-4,22)

```

### Volcano Plot {.tabset .tabset-fade .tabset-pills}
```{r design0.Volcano, echo=TRUE, dpi = 300}
vp = ggplot(data, aes(logFC, P.Value, color = fifteen))
vp + geom_point() +
  scale_color_manual(values = c("black", "red"), labels = c("> 0.05", "< 0.05")) + 
  labs(title = "Gene Level Volcano Plot", color = "FDR") +
  #theme(plot.title = element_text(hjust = 0.5)) +
  theme_classic() +
  xlim(-10,10) + ylim(0, 10) + ylab("-log10 pvalue")

```

## DEG table for download
```{r design0.lists, echo=TRUE}
## Get conversion table for Gencode 30
gencode_30 = read.table("~/pd-omics/katia/ens.geneid.gencode.v30")
colnames(gencode_30) = c("ensembl","symbol")

res$ensembl = rownames(res)
res_name = merge(res, gencode_30, by="ensembl")
rownames(res_name) = res_name$ensembl
res_name = res_name[order(res_name$adj.P.Val), ]
res_name = res_name[, c("symbol", "logFC", "AveExpr", "t", "P.Value", "adj.P.Val", "z.std")]
   
deg_lists = res_name[which(res_name$adj.P.Val<0.05),]
createDT(res_name)

write.table(res_name, file = paste0(work_plots, "sex_dream_3rd.txt"), quote = F, sep = "\t")

```

## Directionality
LogFC > 0 = UP regulated in MALE.  

The pvalues in the boxplots are different from the DEG results. We are using the ggpubr library to compare the means. Test = Wilcoxon. 

```{r direction, echo=TRUE, fig.width=6, fig.height=6, dpi=300}

top_6 = head(res_name)
top_6$ensembl = rownames(top_6)
genes_voom = genes_counts_voom_3rd[rownames(genes_counts_voom_3rd) %in% rownames(genes_counts_nosex),]
#genes_voom = genes_counts_voom[, rownames(metadata4deg)] # Voom 1st pass only for the samples used in this comparison 
gene2check = as.data.frame( genes_voom[rownames(top_6) ,])
gene2check$ensembl = rownames(gene2check)
gene2check = merge(gene2check, top_6[, c("symbol", "ensembl")], by = "ensembl")

gene2check_m = melt(gene2check, id.vars = c("ensembl", "symbol"))
gene2check_charac = merge(gene2check_m, metadata4deg, by.x = "variable", by.y = "donor_tissue")

ggplot(gene2check_charac, aes(x = sex, y = value, fill = sex)) +
  geom_boxplot(notch = T, na.rm = T,  outlier.color = NA) +
 # geom_jitter() +
 # scale_fill_manual(values = c("magenta", "orange")) +
  theme_classic() +
  labs(x = "Gender", y = "Gene expression") +
  ggpubr::stat_compare_means(label = "p.format", label.x.npc = "left", method = "wilcox.test") +
  facet_wrap(~symbol, scales = "free_y")

```

```{r}
sessionInfo()
```

